### ISSUE TITLE
No support for YAML anchors?

### ISSUE THREAD

**Issue by PiCT0** [ARM-software/workload-automation#973]
Consider the following YAML agenda:
```YAML
workloads:
-   name: dhrystone
    params: &dhrystone_single_params
        cleanup_assets: true
        cpus: 0
        delay: 3
        duration: 0
        mloops: 10
        threads: 1
-   name: dhrystone
    params:
        <<: *dhrystone_single_params
        threads: 4
```
which uses YAML [anchor nodes](https://en.wikipedia.org/wiki/YAML#Advanced_components) and [alias nodes](https://yaml.org/spec/1.2/spec.html#alias) to express the redundancy between all `dhrystone` parameters used (except from the overloaded `threads`).

When `run`ning this agenda in WA, I receive:
```
ERROR    Error parsing Agenda <name_of_the_agenda>: Syntax Error on line 11: None
```
However, PyYAML [seems to support aliases](https://pyyaml.org/wiki/PyYAMLDocumentation) so, if it's not coming from my setup, this error could point to a misuse of PyYAML by WA(?) I believe PyYAML should make the "duplication" transparent to the calling code so that WA might not even be aware of the use of YAML aliases.

This is using:
  - WA 3.1.3.dev1 (`pip`-installed in development mode);
  - PyYAML 5.1

**Comment by setrofim** [comment]
The reported syntax error of "None" is confusing due to an issue with our serializer error reporting which is fixed  by https://github.com/ARM-software/workload-automation/pull/974.

The actual error is

```
ERROR    Error parsing Agenda test.yaml: Syntax Error on line 11: could not determine a constructor for the tag 'tag:yaml.org,2002:merge'                                                                                    
ERROR      in "test.yaml", line 12, column 9
```
Will need to look further into why that is happening.

**Comment by marcbonnici** [comment]
This issue seems to lie with constructing pairs for our custom dict constructor (https://github.com/ARM-software/workload-automation/blob/master/wa/utils/serializer.py#L212) when converting mappings back to OrderedDicts. 
If this is removed constructor then the parsing seems to work as expected however not sure why this would be causing that issue.

**Comment by setrofim** [comment]
Should be fixed by: https://github.com/ARM-software/workload-automation/pull/976

**Comment by pietos01** [comment]
While the error message got much better, this issue is not fixed by #976 on my side, running the previous agenda from `master` (b4ea2798), I now get the error message you posted earlier.

From a `git bisect`, it looks like YAML anchors stopped working with 769aae30.

**Comment by setrofim** [comment]
Erm apologies -- I've fucked up the pull requests. That one ended up only containing the error message patch. The actual fix is inside https://github.com/ARM-software/workload-automation/pull/977, and @marcbonnici  has since pushed another update to that. Please could you confirm you're still getting the issue with the latest master?

**Comment by pietos01** [comment]
Haven't I tested #977 by testing b4ea279? Are YAML anchors working for you in #977? If so, which version of PyYAML do you have and, if relevant, of devlib (I'm also on master there)?

**Comment by setrofim** [comment]
>Haven't I tested #977 by testing b4ea279? Are YAML anchors working for you at b4ea279?

Erm, yes if you're at that commit, then you should have the fix.

 >If so, which version of PyYAML do you have and, if relevant, of devlib (I'm also on master there)?

My PyYAML version is 3.12. The `devlib` version is not relevant, as the issue is entirely within WA on configuration handling code (execution is not getting to the point where devlib  first gets invoked).

Probably a silly question, but -- are you sure you're picking up the latest code when you're  WA? Could you stick a break point at  https://github.com/ARM-software/workload-automation/blob/master/wa/utils/serializer.py#L233 and make sure you're hitting it? If that's the case, please could you provided the exact error message you are now getting?

**Comment by marcbonnici** [comment]
Sorry but it turns out that my update ce59318e6678eaae66f060d7bfa51c9676658cea actually did break this again. (Time to add a test for this to prevent this.)
It looks like the original "fix" for this actually solved the issue by accidentally bypassing our custom loader...

**Comment by marcbonnici** [comment]
Ok could you give https://github.com/ARM-software/workload-automation/pull/984 a try and see if that resolves the issue?

**Comment by pietos01** [comment]
I can confirm that this issue is solved as of a6cf62dd (from #984) for the agenda I provide above!
